renormalizer/mps/mp.py

```diff
@@ -645,11 +603,68 @@ def _update_mps(self, cstruct, cidx, qnbigl, qnbigr, Mmax, percent=0):
+                Uset1, SUset1, qnlnew1, Vset1, SVset1, qnrnew1 = svd_qn.svd_qn(
+                    asnumpy(cstruct), qnbigl1, qnbigr1, self.qntot, system=system
+                )
+                qnbigl2, qnbigr2, _ = self._get_big_qn(cidx, swap=True)
+                if cstruct.ndim == 4:
+                    cstruct2 = asnumpy(cstruct).transpose(0, 2, 1, 3)
+                else:
+                    assert cstruct.ndim == 6
+                    cstruct2 = asnumpy(cstruct).transpose(0, 3, 4, 1, 2, 5)
+                if self.compress_config.ofs_swap_jw:
+                    assert cstruct2.ndim == 4
+                    cstruct2 = cstruct2.copy()
+                    cstruct2[:, 1, 1, :] = -cstruct2[:, 1, 1, :]
+                Uset2, SUset2, qnlnew2, Vset2, SVset2, qnrnew2 = svd_qn.svd_qn(
+                    cstruct2, qnbigl2, qnbigr2, self.qntot, system=system
+                )
+                entropy1 = calc_vn_entropy(SUset1**2)
+                entropy2 = calc_vn_entropy(SUset2**2)
+                loss1 = (np.sort(SUset1)[::-1][Mmax:] ** 2).sum()
+                loss2 = (np.sort(SUset2)[::-1][Mmax:] ** 2).sum()
+                ofs = self.compress_config.ofs
+                if ofs is OFS.ofs_d:
+                    should_retain = loss1 <= loss2
+                elif ofs is OFS.ofs_ds:
+                    if loss1 < 1e-10 and loss2 < 1e-10:
+                        # at the end of the chain
+                        should_retain = entropy1 < entropy2
+                    else:
+                        should_retain = loss1 <= loss2
+                elif ofs is OFS.ofs_s:
+                    should_retain = entropy1 < entropy2
+                else:
+                    assert ofs is  OFS.ofs_debug
+                    should_retain = True
+                logger.debug(f"OFS: site index {cidx}, should swap: {not should_retain}, "
+                             f"S: {entropy1}, {entropy2}, loss: {loss1}, {loss2}")
+                if should_retain:
+                    Uset, SUset, qnlnew, Vset, SVset, qnrnew = \
+                        Uset1, SUset1, qnlnew1, Vset1, SVset1, qnrnew1
+                else:
+                    Uset, SUset, qnlnew, Vset, SVset, qnrnew = \
+                        Uset2, SUset2, qnlnew2, Vset2, SVset2, qnrnew2
+                    qnbigl, qnbigr, cstruct = qnbigl2, qnbigr2, cstruct2
+                    new_basis = self.model.basis.copy()
+                    new_basis[cidx[0]:cidx[1] + 1] = reversed(self.model.basis[cidx[0]:cidx[1] + 1])
+                    # previously cached MPOs are destroyed.
+                    # Not sure what is the best way: swap all cached MPOs or simply reconstruct them
+                    # Need some additional testing at production level calculation
+                    self.model: Model = Model(new_basis, self.model.ham_terms, self.model.dipole, self.model.output_ordering)
+                logger.debug(f"DOF ordering: {[b.dof for b in self.model.basis]}")
```

Comment:
What are previously cached MPOs?

Reply:
Operators for the calculation of e and ph occupations and so on.

Reply:
Yes. When swapping happens very frequently, I think it is best to simply reconstruct the MPOs after one sweep. If the swapping rarely happens, then naturally we can just swap the MPOs. I think we need more experience to decide what is the best practice in the general case.